File: packages/utility/comparator.cc
module;
#include <utility>
export module ivm.utility:comparator;

namespace util {

export template <class Type>
struct address_predicate {
	public:
		explicit address_predicate(const Type& value) :
				value_{&value} {}

		auto operator()(const Type& right) const -> bool {
			return value_ == &right;
		}

	private:
		const Type* value_;
};

export template <class Hash>
struct hash_comparator {
		using is_transparent = void;

		template <class Left, class Right>
		auto operator()(const Left& left, const Right& right) const -> bool {
			return Hash{}(left) == Hash{}(right);
		}
};

export template <class Operation, class Fn>
class mapped_comparator {
	public:
		constexpr mapped_comparator(auto&& operation, auto&& fn) :
				operation{std::forward<decltype(operation)>(operation)},
				fn{std::forward<decltype(fn)>(fn)} {}

		constexpr auto operator()(const auto& left, const auto& right) const {
			return operation(fn(left), fn(right));
		}

	private:
		[[no_unique_address]] Operation operation;
		[[no_unique_address]] Fn fn;
};

template <class Operation, class Fn>
mapped_comparator(Operation&&, Fn&&) -> mapped_comparator<Operation, Fn>;

} // namespace util

File: packages/isolated-vm/utility/comparator.ts
/** @internal */
export type Comparator<Type> = (left: Type, right: Type) => number;

/** @internal */
export function mappedComparator<Type, Result>(
	comparator: Comparator<Result>,
	map: (value: Type) => Result,
): Comparator<Type> {
	return (left, right) => comparator(map(left), map(right));
}

/** @internal */
export function selectKey<Key, Value>(entry: [Key, Value]): Key {
	return entry[0];
}

/** @internal */
export function stringComparator(left: string, right: string) {
	return left < right ? -1 : left === right ? 0 : 1;
}
